### PR TITLE
Cross-dataset: 360min multi-seed replication — variance at doubled training budget

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Current best results across TF, AF, and DAM were each achieved with a single seed at 180min timeout. We do not know whether those results are reproducible or luck. This experiment doubles the training budget to 360min and runs 4 seeds per dataset to measure variance and confirm (or refute) that the reported baselines are representative. A secondary goal is to check whether gc=1.0 (gradient clipping) alone, or combined with weight decay (WD=1e-2), further reduces variance at the longer budget.

If variance is low and means are near baseline, we have high confidence in our current best configs. If variance is high, we need to report confidence intervals and/or adopt more stable training procedures.

## Instructions

Use `--wandb_group extended-seeds-360m` on all runs.

**GPU 0 — TF seed 7**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --no-use-ema --epochs 999 --enable-fourier \
  --optimizer lion --model-slices 64 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --seed 7 \
  --wandb_group extended-seeds-360m
```

**GPU 1 — TF seed 42**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --no-use-ema --epochs 999 --enable-fourier \
  --optimizer lion --model-slices 64 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --seed 42 \
  --wandb_group extended-seeds-360m
```

**GPU 2 — TF seed 123**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --no-use-ema --epochs 999 --enable-fourier \
  --optimizer lion --model-slices 64 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --seed 123 \
  --wandb_group extended-seeds-360m
```

**GPU 3 — TF seed 456**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --no-use-ema --epochs 999 --enable-fourier \
  --optimizer lion --model-slices 64 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --seed 456 \
  --wandb_group extended-seeds-360m
```

**GPU 4 — AF seed 7**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --no-use-ema --epochs 999 --enable-fourier \
  --seed 7 \
  --wandb_group extended-seeds-360m
```

**GPU 5 — AF seed 42**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --no-use-ema --epochs 999 --enable-fourier \
  --seed 42 \
  --wandb_group extended-seeds-360m
```

**GPU 6 — AF gc=1.0 seed 42**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --no-use-ema --epochs 999 --enable-fourier \
  --grad-clip 1.0 \
  --seed 42 \
  --wandb_group extended-seeds-360m
```

**GPU 7 — AF gc=1.0 + WD=1e-2 seed 42**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --no-use-ema --epochs 999 --enable-fourier \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --seed 42 \
  --wandb_group extended-seeds-360m
```

Report mean ± std of `val_primary/surface_pressure_mae` across TF seeds and `val_primary/surface_mse` across AF seeds. Compare AF gc/WD variants directly to the AF seed-42 baseline run on GPU 5.

## Baseline

- TF: `val_primary/surface_pressure_mae` = 30.10 (PR #2810)
- AF: `val_primary/surface_mse` = 0.001236 (PR #2828)
- DAM: `val_primary/surface_rel_l2_pct` = 4.619% (PR #2691)

Reproduce TF baseline:
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --no-use-ema --epochs 999 --enable-fourier \
  --optimizer lion --model-slices 64 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition
```

Reproduce AF baseline:
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --no-use-ema --epochs 999 --enable-fourier
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT vegeta:

## Note: Corrected commands to match baselines

The GPU commands in the PR body are missing several critical hyperparameters that define the current baselines. Running the commands as-written would use defaults (lr=3e-4, T_max=150, no gc, Lion optimizer) which are far from the current best configs.

**TF corrections (GPUs 0-3):** Added `--lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2` to match PR #2810 baseline (30.10 MAE).

**AF corrections (GPUs 4-7):** Added `--optimizer adamw --lr 7e-4 --cosine-t-max 5 --model-layers 2 --model-hidden-dim 256 --model-heads 4` to match the 2L/256d architecture from PR #2828. GPUs 4-5 run WITHOUT gc/WD (ablation), GPU 6 adds gc=1.0, GPU 7 adds gc=1.0 + WD=1e-2 — preserving the original experiment design intent.

Proceeding with corrected commands. Let me know if you intended a different config.

**Comment by tcapelle** [comment]
STUDENT vegeta:

## Bug fix: TF runs restarted with correct weight-decay

**Timestamp:** 2026-04-21 21:02 UTC

Discovered that the TF runs (GPUs 0-3) were launched **without** `--weight-decay 1e-2`, using the default WD=1e-4 instead. The BASELINE.md reproduce command for PR #2810 clearly specifies WD=1e-2.

Killed the 4 TF runs (only at epoch 1, ~16 min wasted) and relaunched with the correct config:
```
--lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2
```

AF runs (GPUs 4-7) are unaffected and continuing at epoch ~25. The AF ablation design is intentional: GPUs 4-5 run without gc/WD, GPU 6 adds gc=1.0, GPU 7 adds gc=1.0+WD=1e-2 (full baseline match).

W&B run names now use `-v2` suffix for the TF restarts.

**Comment by tcapelle** [comment]
STUDENT vegeta:

## Bug fix: TF runs restarted (v3) — persistent_workers deadlock

**Timestamp:** 2026-04-21 21:32 UTC

All 4 TF runs (GPUs 0-3) deadlocked after epoch 1. Symptoms:
- 0% CPU and 0% GPU utilization despite processes still being alive
- Log files stopped updating ~12 min after epoch 1 completed
- Processes in sleeping state with no progress

**Root cause:** Likely a PyTorch DataLoader `persistent_workers=True` deadlock (known issue with fork-based multiprocessing and large datasets). AF runs on GPUs 4-7 are unaffected.

**Fix:** Killed stuck processes and relaunched all 4 TF runs with `--no-persistent-workers`. Run names now use `-v3` suffix.

AF runs continue unaffected at epoch ~90, progressing normally.

**Comment by tcapelle** [comment]
STUDENT vegeta:

## Results

### TandemFoil — 4-seed replication (360 min budget)

| Seed | Epochs | Best `val_primary/surface_pressure_mae` | Best Epoch | W&B Run |
|------|--------|---------------------------------------|------------|---------|
| 7    | 36     | 45.14                                 | 35         | [n0wuslwo](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/n0wuslwo) |
| 42   | 37     | 41.42                                 | 37         | [w883w8vi](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/w883w8vi) |
| 123  | 37     | 43.39                                 | 31         | [p08c8yhh](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/p08c8yhh) |
| 456  | 37     | 43.44                                 | 36         | [86vts2rc](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/86vts2rc) |

**Mean ± Std:** 43.35 ± 1.52 | **Baseline:** 30.10 (ep 157, 180 min)

**Config:** `--dataset tandemfoil --no-use-ema --epochs 999 --enable-fourier --optimizer lion --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-persistent-workers --wandb-group extended-seeds-360m`

**Peak memory:** 34–54 GiB (varied by seed; GPU0=54 GiB, GPUs 1-3=34-36 GiB)

⚠️ **Critical caveat:** TF runs only reached 36–37 epochs in 360 min (~9.7 min/epoch) vs the baseline's 157 epochs in 180 min (~1.15 min/epoch). This 8.4x slowdown is caused by:
1. `--no-persistent-workers` — required to avoid a DataLoader deadlock (persistent_workers=True deadlocked all TF runs after epoch 1, see bug fix comments above). Worker respawning adds significant per-epoch overhead.
2. Running 4 TF + 4 AF processes simultaneously likely caused I/O and CPU contention for data loading.

**TF results are NOT comparable to baseline** — the runs did not get enough epochs to converge. The variance measurement (±1.52) is from an under-converged regime and does not represent final performance variance.

---

### AirfRANS — seed variance + gc/wd ablation (360 min budget)

| Config | Seed | Epochs | Best `val_primary/surface_mse` | Best Epoch | W&B Run | Notes |
|--------|------|--------|-------------------------------|------------|---------|-------|
| No gc/wd | 7  | 756    | 0.001198                      | 542        | [3r0g6u3y](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3r0g6u3y) | Beats baseline (−3%) |
| No gc/wd | 42 | 760    | 0.002477                      | 308        | [nbc25ot7](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/nbc25ot7) | **Diverged to NaN at ep607** (154 NaN epochs) |
| gc=1.0 | 42  | 754    | **0.000598**                  | 517        | [d7a0z1hk](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/d7a0z1hk) | **52% below baseline** |
| gc=1.0 + wd=1e-2 | 42 | 759 | **0.000694**             | 718        | [nvllyhmf](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/nvllyhmf) | **44% below baseline** |

**Baseline:** 0.001236 (ep 349, 180 min, gc=1.0 + wd=1e-2, 2L/256d) | **External target:** 0.0043

**AF Config (all runs):** `--dataset airfrans --no-use-ema --epochs 999 --enable-fourier --optimizer adamw --lr 7e-4 --cosine-t-max 5 --model-layers 2 --model-hidden-dim 256 --model-heads 4 --wandb-group extended-seeds-360m`

Additional flags per variant:
- GPUs 4-5 (no gc/wd): defaults (no `--grad-clip`, `--weight-decay` = 1e-4 default)
- GPU 6 (gc=1.0): `--grad-clip 1.0`
- GPU 7 (gc=1.0 + wd=1e-2): `--grad-clip 1.0 --weight-decay 1e-2`

**Peak memory:** 24–29 GiB

---

### Key Findings

**1. gc=1.0 dramatically improves AF convergence and stability (same seed 42 comparison):**
- Without gc: best 0.002477, diverged to NaN at ep607
- gc=1.0: best **0.000598** (4.1x better), fully stable through 754 epochs
- gc=1.0 + wd=1e-2: best **0.000694** (3.6x better), fully stable through 759 epochs

This is a controlled ablation (same seed, same config, only gc/wd differs). gc=1.0 prevents the gradient explosion that caused the no-gc run to diverge, and enables the optimizer to find much deeper basins.

**2. AF seed variance is HIGH without gc:**
- Seed 7 (no gc): 0.001198 — beats baseline
- Seed 42 (no gc): 0.002477 — 2x above baseline, then diverges
- Range: 2.07x between best and worst seed

**3. gc=1.0 alone outperforms gc=1.0 + wd=1e-2:**
- gc=1.0: 0.000598 (best at ep517)
- gc=1.0+wd=1e-2: 0.000694 (best at ep718)
- The weight decay may be slightly over-regularizing, though both massively beat baseline.
- Note: the AF baseline (0.001236) was achieved WITH gc=1.0 + wd=1e-2, so gc=1.0 alone at doubled budget is 52% better.

**4. Doubled training budget (360 vs 180 min) is highly beneficial for AF:**
- Baseline: 0.001236 at ep349 (180 min)
- gc=1.0 360 min: 0.000598 at ep517 — the extra epochs unlock significantly deeper basins
- Even without gc, seed 7 (0.001198 at ep542) slightly beats the 180-min baseline

**5. TF results are inconclusive due to persistent_workers deadlock:**
- The deadlock (all 4 runs frozen after epoch 1) required `--no-persistent-workers` which caused an 8.4x slowdown
- Only 36-37 epochs completed vs the baseline's 157 — insufficient for convergence
- The low variance (±1.52) at this early stage suggests seeds are similar in early training

---

### What happened

**AF:** The doubled training budget combined with gc=1.0 produced remarkable results — 52% below the existing best baseline. The gc ablation provides clear evidence that gradient clipping is essential for AirfRANS training stability, preventing divergence and enabling deeper optimization. The no-gc seed42 diverging to NaN while gc=1.0 seed42 achieves the best result is a particularly clean demonstration.

**TF:** A PyTorch DataLoader `persistent_workers=True` deadlock caused all 4 runs to freeze after epoch 1. The workaround (`--no-persistent-workers`) prevented deadlock but caused severe slowdown. The TF variance question remains unanswered — these runs need to be re-run with persistent workers on a single-GPU-per-run setup to get comparable epoch counts.

### Suggested follow-ups

1. **Re-run TF seeds individually** (one GPU at a time) with `persistent_workers=True` to get proper epoch counts and resolve the variance question.
2. **AF gc=1.0 is a new baseline candidate** at 0.000598 — this is 52% below the current best and could be the new AF SOTA.
3. **Try gc=1.0 without wd** at 180-min budget to see if the gc-only advantage holds at standard budget.
4. **Investigate the persistent_workers deadlock** — may be a broader issue affecting TandemFoil training in general (only TF was affected, not AF).

**Comment by tcapelle** [comment]
**ADVISOR REVIEW — PR #2906**

Thank you for the thorough multi-seed ablation at 360 min. The AirfRANS findings here are genuinely important. The TandemFoil results, unfortunately, are not usable — I'll explain what went wrong and what to do about it.

---

## AirfRANS — ACCEPTED ✓ (New SOTA)

The gc=1.0 seed=42 run achieves **val MSE = 0.000598** at epoch 517 (W&B: d7a0z1hk). This beats the previous best of 0.000627 (PR #2902) by **4.6%** — a meaningful improvement. I'm updating the BASELINE accordingly.

Key insights from this ablation:

1. **gc=1.0 is essential for AirfRANS stability.** Without it, seed=42 hit NaN at epoch 607 (run nbc25ot7, val=0.002477). Seed=7 barely held on (0.001198). With gc=1.0, the run stays stable and finds a much deeper basin.

2. **gc=1.0 alone beats gc=1.0 + WD=1e-2.** The combined config (nvllyhmf) settled at 0.000694 vs the gc-only config's 0.000598. This is a surprising finding — WD was previously believed necessary for AF, but the data challenges that assumption. WD may be limiting the model's ability to reach very low-loss configurations.

3. **360 min budget is highly beneficial for AirfRANS.** 0.000598 at 360 min vs 0.000627 at 180 min with WD — the model is still improving at the 180-min mark and benefits from continued training.

The AF data is merged into the research record. The new AF baseline is 0.000598.

---

## TandemFoil — REJECTED ✗ (Invalid — persistent_workers deadlock)

The TF results are **not usable**. All 4 seeds only completed 36–37 epochs in 360 minutes due to the persistent_workers deadlock. At 8.4× slowdown, you were effectively running at ~4 epochs/hour instead of ~33/hour. The mean val MAE of 43.35 is meaningless — we need hundreds of epochs to see real TF performance.

The deadlock occurs when concurrent processes fight over DataLoader worker pools with `persistent_workers=True`. The workaround `--no-persistent-workers` kills throughput.

**For the TF redo, please follow these specific instructions:**

1. Run **one TF process at a time** (no concurrent TF+AF on the same node — they share DataLoader workers and cause the deadlock)
2. Do NOT use `--no-persistent-workers` — the default (`persistent_workers=True`) is correct and needed for speed
3. Run 4 seeds (7, 42, 123, 456) **sequentially or on separate isolated nodes**, not concurrently
4. Use the current best TF config: `--dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999`
5. Use `--wandb_group vegeta-tf-360min-redo` so runs are grouped together
6. Budget: 360 min per seed

Goal: establish true seed variance at 360 min for the TF baseline config.

---

Sending this back for the TF redo. The AF contribution stands and is already incorporated.

**Comment by tcapelle** [comment]
ADVISOR: AF results accepted — new SOTA 0.000598 (W&B d7a0z1hk). TF results invalid due to persistent_workers deadlock (only 36-37 epochs in 360 min). Please redo TF with single isolated process per seed — see advisor review comment above for exact instructions.

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
